### ISSUE TITLE
Miles SFT does not seem to work for VLM Qwen3-VL-8B-Instruct (using Megatron as backend)

### ISSUE THREAD

**Issue by deepak-vij** [radixark/miles#650]
I am doing SFT training for the following miles repo example. The training runs successfully on B200. I could see tensor/loss flattening out after 20 epochs withinin Tensorboard. But when I try to access the fine-tuned model in SGLang, the model output is all gibberish. I also had to modify the sft_rollout.py python code (see below) in order to enable   --apply-chat-template training argument. I am successfully able to access the baseline model (Qwen3-VL-8B-Instruct) successfully in SGLang.

@claude @fzyzcjy @Lyken17 @Ying1123 @yushengsu-thu, it would be great to get pointers as I am trying ensure e2e SFT process works and I can access the fine-tuned model in SGLang. Thanks in anticipation.

https://github.com/radixark/miles/blob/main/examples/geo3k_vlm/README.md
++++++++++++++++++++++++++ sft_rollout.py +++++++++++++++++
    for i, sample in enumerate(samples):
        (sample,) = sample
        # messages = sample.prompt

       # Try to get data from sample attributes, then from metadata dictionary
        problem_text = getattr(sample, 'problem', sample.metadata.get('problem', ''))
        answer_text = getattr(sample, 'answer', sample.metadata.get('answer', ''))
        images_list = getattr(sample, 'images', sample.metadata.get('images', []))

        # Reconstruct the structured messages
        messages = [
        {
            "role": "user",
            "content": [
                {"type": "image", "image": images_list[0] if images_list else None},
                {"type": "text", "text": problem_text}
            ]
        },
        {
            "role": "assistant", 
            "content": [{"type": "text", "text": answer_text}]
        }
        ]
++++++++++++++++++++++++++ Training Script ++++++++++++++++
TRAIN_BACKEND=${MILES_SCRIPT_TRAIN_BACKEND:-"megatron"}
MODEL_NAME=${MILES_SCRIPT_MODEL_NAME:-"Qwen3-VL-8B-Instruct"}
DATASET_NAME=${MILES_SCRIPT_DATASET_NAME:-"chenhegu/geo3k_imgurl"}
NUM_GPUS=${MILES_SCRIPT_NUM_GPUS:-8}
DATASET_LOCAL_NAME=$(basename "$DATASET_NAME")

VALID_MODELS="
  Qwen3-VL-2B-Instruct
  Qwen3-VL-4B-Instruct
  Qwen3-VL-8B-Instruct
  Qwen3-VL-2B-Thinking
  Qwen3-VL-4B-Thinking
  Qwen3-VL-8B-Thinking
  Qwen3-VL-30B-A3B-Instruct
  Qwen3-VL-235B-A22B-Instruct
  Qwen3-VL-30B-A3B-Thinking
  Qwen3-VL-235B-A22B-Thinking
"
if ! echo "$VALID_MODELS" | grep -qw "$MODEL_NAME"; then
   echo "Error: MODEL_NAME must be one of: $VALID_MODELS"
   exit 1
fi

MODEL_NAME_LOWER=$(echo "$MODEL_NAME" | tr '[:upper:]' '[:lower:]')

if [ -z "$MILES_SCRIPT_EXTERNAL_RAY" ] || [ "$MILES_SCRIPT_EXTERNAL_RAY" = "0" ]; then
   USE_EXTERNAL_RAY=0
else
   USE_EXTERNAL_RAY=1
fi

pkill -9 sglang
sleep 3
if [ "$USE_EXTERNAL_RAY" = "0" ]; then
   ray stop --force
   pkill -9 ray
fi
pkill -9 miles
sleep 3
if [ "$USE_EXTERNAL_RAY" = "0" ]; then
   pkill -9 ray
fi
pkill -9 miles
pkill -9 redis

set -ex

export PYTHONBUFFERED=16

NVLINK_COUNT=$(nvidia-smi topo -m 2>/dev/null | grep -o 'NV[0-9][0-9]*' | wc -l)
if [ "$NVLINK_COUNT" -gt 0 ]; then
   HAS_NVLINK=1
else
   HAS_NVLINK=0
fi
echo "HAS_NVLINK: $HAS_NVLINK (detected $NVLINK_COUNT NVLink references)"

mkdir -p /root/models /root/datasets
if [ ! -d "/root/models/${MODEL_NAME}" ]; then
   hf download Qwen/${MODEL_NAME} --local-dir /root/models/${MODEL_NAME}
fi
if [ ! -d "/root/datasets/${DATASET_LOCAL_NAME}" ]; then
   hf download --repo-type dataset ${DATASET_NAME} --local-dir /root/datasets/${DATASET_LOCAL_NAME}
fi

CKPT_ARGS=(
   --hf-checkpoint /root/models/${MODEL_NAME}
   --load /root/models/${MODEL_NAME}
)

SFT_ARGS=(
   --rollout-function-path miles.rollout.sft_rollout.generate_rollout
   --prompt-data /root/datasets/${DATASET_LOCAL_NAME}/train_formatted_100.parquet
   --input-key messages
   --apply-chat-template
   --rollout-shuffle
   --num-epoch 30
   --rollout-batch-size 96
   --transformer-impl transformer_engine
 
   --num-steps-per-rollout 1

   --loss-type sft_loss
   --calculate-per-token-loss
   --disable-compute-advantages-and-returns
   --debug-train-only
   --tensor-model-parallel-size 4
   --save /data05/deepak/eval
   --save-interval 50
   --no-save-optim
   --save-hf /data05/deepak/final_model
)

MULTIMODAL_KEYS='{"image": "images"}'

OPTIMIZER_ARGS=(
   --optimizer adam
   --lr 1e-5
   --lr-decay-style cosine
   --min-lr 1e-6
   --lr-warmup-fraction 0.1
   --weight-decay 0.1
   --adam-beta1 0.9
   --adam-beta2 0.95
)

if [ -n "$WANDB_API_KEY" ]; then
    WANDB_ARGS=(
        --use-wandb
        --wandb-project miles-geo3k-vlm-sft
        --wandb-group ${MODEL_NAME_LOWER}-${TRAIN_BACKEND}
        --wandb-key ${WANDB_API_KEY}
        --disable-wandb-random-suffix
    )
else
    WANDB_ARGS=(
      --use-tensorboard
    )
fi

if [ "$TRAIN_BACKEND" = "fsdp" ]; then
    BACKEND_ARGS=(
      --train-backend fsdp
      --gradient-checkpointing
      --attn-implementation flash_attention_3
      --update-weight-buffer-size 536870912
    )
else
    BACKEND_ARGS=(
      --train-backend megatron
      --tensor-model-parallel-size 4
      --sequence-parallel
      --pipeline-model-parallel-size 1
      --context-parallel-size 1
      --expert-model-parallel-size 1
      --expert-tensor-parallel-size 1
      --recompute-method uniform
      --recompute-num-layers 1
      --use-dynamic-batch-size
      --max-tokens-per-gpu 4096
      --attention-dropout 0.0
      --hidden-dropout 0.0
      --accumulate-allreduce-grads-in-fp32
      --attention-softmax-in-fp32
      --attention-backend flash
      --megatron-to-hf-mode bridge
    )

   MILES_DIR="$(cd -- "$(dirname -- "${BASH_SOURCE[0]}")/../.." &>/dev/null && pwd)"
   MODEL_ARGS_FILE=$(echo "$MODEL_NAME" | sed 's/-Instruct//g; s/-Thinking//g; s/Qwen3-VL-/qwen3-/g; s/-2B/-1.7B/g')
   MODEL_ARGS_ROTARY_BASE=5000000 source "${MILES_DIR}/scripts/models/${MODEL_ARGS_FILE}.sh"
fi

if [ "$USE_EXTERNAL_RAY" = "0" ]; then
   export MASTER_ADDR=${MASTER_ADDR:-"127.0.0.1"}
   export no_proxy="127.0.0.1,${MASTER_ADDR}"
   ray start --head --node-ip-address ${MASTER_ADDR} --num-gpus ${NUM_GPUS} --disable-usage-stats --dashboard-host=0.0.0.0 --dashboard-port=8265
fi

RUNTIME_ENV_JSON="{
  \"env_vars\": {
    \"PYTHONPATH\": \"/root/deepak/Megatron-LM/:/root/miles/custom_hooks/\",
    \"CUDA_DEVICE_MAX_CONNECTIONS\": \"1\",
    \"NCCL_NVLS_ENABLE\": \"${HAS_NVLINK}\"
  }
}"

ray job submit --address="http://127.0.0.1:8265" \
   --runtime-env-json="${RUNTIME_ENV_JSON}" \
   -- python3 train_async.py \
   --actor-num-nodes 1 \
   --actor-num-gpus-per-node ${NUM_GPUS} \
   --multimodal-keys "${MULTIMODAL_KEYS}" \
   ${MODEL_ARGS[@]} \
   ${CKPT_ARGS[@]} \
   ${SFT_ARGS[@]} \
   ${EVAL_ARGS[@]} \
   ${OPTIMIZER_ARGS[@]} \
   ${WANDB_ARGS[@]} \
   ${BACKEND_ARGS[@]}
